### PR TITLE
Add logging to 'business' events

### LINKED ISSUE #590: Add logging for "business events"
Music Store currently does no logging of its own.

To better demonstrate logging, it should log business events such as:

* Adding/removing items to/from the cart
* Starting checkout
* Completed checkout
* User register/login/logout activity (success + failure)

I recommend using the regular `ILogger<T>` pattern for logging (not the fancy lambda stuff).

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #590.

### CODE REVIEW

File: src/MusicStore/Controllers/AccountController.cs [hunk 1]
```diff
@@ -8,19 +8,24 @@
...
     [Authorize]
     public class AccountController : Controller
     {
+        private ILogger _logger { get; set; }
+
         public AccountController(
             UserManager<ApplicationUser> userManager,
-            SignInManager<ApplicationUser> signInManager)
+            SignInManager<ApplicationUser> signInManager,
+            ILogger logger)
```
Comment (kichalla) [inline src/MusicStore/Controllers/AccountController.cs]: `ILogger<AccountController>` and similarly in other files

File: src/MusicStore/Startup.cs [hunk 2]
```diff
@@ -81,12 +83,15 @@ public void ConfigureServices(IServiceCollection services)
             // Add the system clock service
             services.AddSingleton<ISystemClock, SystemClock>();
 
+            services.AddSingleton<ILogger>((s) => { return new ConsoleLogger("main", null, false); });
```
Comment (kichalla) [inline src/MusicStore/Startup.cs]: Need not do this. There is already a service which resolves `ILogger<>`
https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs#L28

File: src/MusicStore/Controllers/ShoppingCartController.cs [hunk 3]
```diff
@@ -49,6 +53,7 @@ public ShoppingCartController(MusicStoreContext dbContext)
             cart.AddToCart(addedAlbum);
 
             await DbContext.SaveChangesAsync(requestAborted);
+            _logger.LogInformation("Album {id} was added to the cart.", addedAlbum.AlbumId);
```
Comment (kichalla) [inline src/MusicStore/Controllers/ShoppingCartController.cs]: nit: id -> albumId

File: src/MusicStore/Controllers/AccountController.cs [hunk 4]
```diff
@@ -65,6 +71,7 @@ public IActionResult Login(string returnUrl = null)
             }
             else
             {
+                _logger.LogInformation("Failed to login {userName}.", model.Email);
```
Comment (kichalla) [inline src/MusicStore/Controllers/AccountController.cs]: Warning?

**Comment by kichalla** [comment]
:watch:

File: src/MusicStore/Controllers/AccountController.cs (hunk 4 above)
Comment (ryanbrandenburg) [inline src/MusicStore/Controllers/AccountController.cs]: Eh, depends on your philosophy. If you're trying to surface these to a user in some way then yes, totally a warning. But if these are for debugging and tracing then I don't think someone forgetting their password counts as a warning.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/MusicStore/Controllers/AccountController.cs [hunk 5]
```diff
@@ -8,19 +8,24 @@
...
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
 using MusicStore.Models;
 
 namespace MusicStore.Controllers
 {
     [Authorize]
     public class AccountController : Controller
     {
+        private ILogger<AccountController> _logger { get; set; }
```
Comment (kichalla) [inline src/MusicStore/Controllers/AccountController.cs]: Make all these readonly fields

File: src/MusicStore/Controllers/CheckoutController.cs [hunk 6]
```diff
@@ -78,17 +88,24 @@ public IActionResult AddressAndPayment()
...
+
             // Validate customer owns this order
             bool isValid = await dbContext.Orders.AnyAsync(
                 o => o.OrderId == id &&
-                o.Username == HttpContext.User.Identity.Name);
+                o.Username == userName);
 
             if (isValid)
             {
+                _logger.LogInformation("User {userName} completed checkout on order {orderId}.", userName, id);
```
Comment (kichalla) [inline src/MusicStore/Controllers/CheckoutController.cs]: @Eilon Should these use the `Logger.Define` approach or is this ok?

Comment (ryanbrandenburg) [inline src/MusicStore/Controllers/CheckoutController.cs]: You mean [these](https://github.com/aspnet/Mvc/blob/a78f77afde003c4a3fcf5dd7b6dc13dd9c85f825/src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/MvcJsonLoggerExtensions.cs#L17)? He specifically said not to use those to keep it simple.

Comment (kichalla) [inline src/MusicStore/Controllers/CheckoutController.cs]: Yes, I meant them. Sure :+1:

**Comment by kichalla** [comment]
:shipit: after addressing the minor comment

File: src/MusicStore/Controllers/AccountController.cs (hunk 4 above)
Comment (Eilon) [inline src/MusicStore/Controllers/AccountController.cs]: I agree with @kichalla , from a business perspective this is a warning - it's something that could raise a red flag during an audit of logs.

Comment (Eilon) [inline src/MusicStore/Controllers/AccountController.cs]: Plug, `login` is a noun, but `log in` is a verb phrase, and I think you want the verb.

**Pull request merged**